### PR TITLE
[PHPMD] Add phpmd rule with ignoring of unused required parameters in the plugins

### LINKED ISSUE #37914: [Issue] [PHPMD] Add phpmd rule with ignoring of unused required parameters in the plugins
This issue is automatically created based on existing pull request: magento/magento2#33918: [PHPMD] Add phpmd rule with ignoring of unused required parameters in the plugins

---------

### Description (*)

This PR improve phpmd `UnusedFormalParameter` rule with ignoring of required unused parameters in the plugins.

- `around` and `after` plugins has 2 required params `$subject` and `$proceed` (or `$result` for `after` plugin).
- `before` plugin has 1 required param `$subject`.

This PR override standard `UnusedFormalParameter` rule and allow don't use `@SuppressWarnings(PHPMD.UnusedFormalParameter)` in the plugins.

### Manual testing scenarios (*)

1. Configure phpmd
2. Remove `@SuppressWarnings(PHPMD.UnusedFormalParameter)` from some plugin
3. Add `PluginInterface` implementation into the class
4. Run phpmd. Expected no `UnusedFormalParameter` errors

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] README.md files for modified modules are updated and included in the pull request if any [README.md predefined sections](https://github.com/magento/devdocs/wiki/Magento-module-README.md) require an update
 - [ ] All automated tests passed successfully (all builds are green)

### PR DESCRIPTION
**Pull request by Den4ik**
### Description (*)

This PR improve phpmd `UnusedFormalParameter` rule with ignoring of required unused parameters in the plugins.

- `around` and `after` plugins has 2 required params `$subject` and `$proceed` (or `$result` for `after` plugin).
- `before` plugin has 1 required param `$subject`.

This PR override standard `UnusedFormalParameter` rule and allow don't use `@SuppressWarnings(PHPMD.UnusedFormalParameter)` in the plugins.

### Manual testing scenarios (*)

1. Configure phpmd
2. Remove `@SuppressWarnings(PHPMD.UnusedFormalParameter)` from some plugin
3. Add `PluginInterface` implementation into the class
4. Run phpmd. Expected no `UnusedFormalParameter` errors

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] README.md files for modified modules are updated and included in the pull request if any [README.md predefined sections](https://github.com/magento/devdocs/wiki/Magento-module-README.md) require an update
 - [ ] All automated tests passed successfully (all builds are green)

### Resolved issues:
1. [x] resolves magento/magento2#37914: [PHPMD] Add phpmd rule with ignoring of unused required parameters in the plugins

### CODE REVIEW

File: dev/tests/static/framework/Magento/CodeMessDetector/Rule/UnusedCode/UnusedFormalParameter.php
Comment (ihor-sviziev) [inline dev/tests/static/framework/Magento/CodeMessDetector/Rule/UnusedCode/UnusedFormalParameter.php]: I think introducing a special interface to fix the static test failures is the wrong way to do it. Can we just check if the class is located in the `*\Plugin\*` namespace?

**Review by ihor-sviziev — CHANGES_REQUESTED** [review]

File: dev/tests/static/framework/Magento/CodeMessDetector/Rule/UnusedCode/UnusedFormalParameter.php
Comment (ihor-sviziev) [reply dev/tests/static/framework/Magento/CodeMessDetector/Rule/UnusedCode/UnusedFormalParameter.php]: Could you also cover your change with unit test? 
Example you can see here:
https://github.com/phpmd/phpmd/tree/master/src/test/php/PHPMD/Rule/CleanCode

Comment (Den4ik) [reply dev/tests/static/framework/Magento/CodeMessDetector/Rule/UnusedCode/UnusedFormalParameter.php]: > I think introducing a special interface to fix the static test failures is the wrong way to do it. Can we just check if the class is located in the *\Plugin\* namespace?

I thought about it, but in fact we could have non plugin classes in `Plugin` namespace.

Comment (Den4ik) [reply dev/tests/static/framework/Magento/CodeMessDetector/Rule/UnusedCode/UnusedFormalParameter.php]: > Could you also cover your change with unit test?

Yep if idea will be approved by core team

Comment (bgorski) [reply dev/tests/static/framework/Magento/CodeMessDetector/Rule/UnusedCode/UnusedFormalParameter.php]: Honestly, while I totally agree with the idea of phpmd skipping those parameters in the check, I also think that introducing a new marker interface for this is a wrong approach. 

While I don't have any experience in writing or improving phpmd checks myself, I believe that you should be able to parse di.xml files from the same module to determine if a class is a plugin or not. Am I correct with this assumption?

Effort-wise, with a marker interface, you just trade the necessity of writing `SuppressWarnings(PHPMD.UnusedFormalParameter)` for the necessity of writing `implements \Magento\Framework\Interception\PluginInterface`.

Comment (Den4ik) [reply dev/tests/static/framework/Magento/CodeMessDetector/Rule/UnusedCode/UnusedFormalParameter.php]: @bgorski I thought about parsing of di.xml but it's very expensive only for tests

Comment (Den4ik) [reply dev/tests/static/framework/Magento/CodeMessDetector/Rule/UnusedCode/UnusedFormalParameter.php]: I'll try to implement parsing and check preformance

Comment (ihor-sviziev) [reply dev/tests/static/framework/Magento/CodeMessDetector/Rule/UnusedCode/UnusedFormalParameter.php]: I think checking the namespace will be a good enough solution. The edge case, when the plugin not in the `Plugin` namespace still could be ignored

File: dev/tests/static/framework/Magento/CodeMessDetector/Rule/UnusedCode/UnusedFormalParameter.php
Comment (ihor-sviziev) [inline dev/tests/static/framework/Magento/CodeMessDetector/Rule/UnusedCode/UnusedFormalParameter.php]: Looks like this class left from previous iteration and doesn't exist 
```suggestion
```

File: dev/tests/static/framework/Magento/CodeMessDetector/Rule/UnusedCode/UnusedFormalParameter.php
Comment (ihor-sviziev) [inline dev/tests/static/framework/Magento/CodeMessDetector/Rule/UnusedCode/UnusedFormalParameter.php]: I think there it would be better to revert checks and add `return`.

File: dev/tests/static/framework/Magento/CodeMessDetector/Rule/UnusedCode/UnusedFormalParameter.php
Comment (ihor-sviziev) [inline dev/tests/static/framework/Magento/CodeMessDetector/Rule/UnusedCode/UnusedFormalParameter.php]: same here

**Review by ihor-sviziev — CHANGES_REQUESTED** [review]
Please cover changes with some unit test

**Comment by Den4ik** [comment]
@ihor-sviziev I'll work on tests in a week

**Comment by Den4ik** [comment]
@magento run all tests

**Comment by Den4ik** [comment]
@ihor-sviziev @bgorski I updated PR with some refactoring and added unit tests. Please approve when all tests will be green and you don't have any suggestions.

**Comment by Den4ik** [comment]
@magento run all tests

**Comment by Den4ik** [comment]
@magento run Functional Tests EE, Functional Tests CE

**Comment by Den4ik** [comment]
@magento run Functinal Tests B2B

**Comment by ihor-sviziev** [comment]
@magento run Functinal Tests B2B

**Comment by Den4ik** [comment]
@magento run Functional Tests B2B

File: dev/tests/static/framework/Magento/CodeMessDetector/Test/Unit/Rule/UnusedCode/UnusedFormalParameterTest.php
Comment (ihor-sviziev) [inline dev/tests/static/framework/Magento/CodeMessDetector/Test/Unit/Rule/UnusedCode/UnusedFormalParameterTest.php]: I think it could be removed

**Review by ihor-sviziev — CHANGES_REQUESTED** [review]

File: dev/tests/static/framework/Magento/CodeMessDetector/Rule/UnusedCode/UnusedFormalParameter.php
Comment (ihor-sviziev) [inline dev/tests/static/framework/Magento/CodeMessDetector/Rule/UnusedCode/UnusedFormalParameter.php:80]: Interesting... why it's named as getImage? 😃

Comment (Den4ik) [reply dev/tests/static/framework/Magento/CodeMessDetector/Rule/UnusedCode/UnusedFormalParameter.php:80]: Both `getName` and `getImage` contain `return $this->name`.
Tests in phpmd source use `getImage`

File: dev/tests/static/framework/Magento/CodeMessDetector/Test/Unit/Rule/UnusedCode/UnusedFormalParameterTest.php
Comment (Den4ik) [reply dev/tests/static/framework/Magento/CodeMessDetector/Test/Unit/Rule/UnusedCode/UnusedFormalParameterTest.php]: Oh, yep

**Comment by Den4ik** [comment]
@magento run all tests

File: dev/tests/static/framework/Magento/CodeMessDetector/Rule/UnusedCode/UnusedFormalParameter.php
Comment (ihor-sviziev) [reply dev/tests/static/framework/Magento/CodeMessDetector/Rule/UnusedCode/UnusedFormalParameter.php:80]: Could you add this as a comment?

**Comment by Den4ik** [comment]
@magento run all tests

**Comment by Den4ik** [comment]
@magento run all tests

**Comment by Den4ik** [comment]
@magento run Functional Tests EE, Functional Tests B2B

**Comment by Den4ik** [comment]
@magento run Functional Tests B2B

**Review by ihor-sviziev — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @ihor-sviziev, thank you for the review.
[ENGCOM-9223](https://jira.corp.magento.com/browse/ENGCOM-9223) has been created to process this Pull Request

**Comment by Den4ik** [comment]
@sidolov @gabrieldagama Could you please check and assign priority to this PR.
I believe that we can assign High Priority and merge it because it's really improve developer experience

**Review by bgorski — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @bgorski, thank you for the review.
[ENGCOM-9223](https://jira.corp.magento.com/browse/ENGCOM-9223) has been created to process this Pull Request

**Comment by Den4ik** [comment]
@magento run all tests

**Comment by Den4ik** [comment]
@magento run all tests

**Comment by engcom-Echo** [comment]
@magento run all tests

**Comment by engcom-Echo** [comment]
@magento run Unit Tests ,Functional Tests CE,Functional Tests EE,Functional Tests B2B

**Comment by engcom-Echo** [comment]
@magento create issue

**Comment by engcom-Echo** [comment]
### :heavy_check_mark: QA Passed

**Manual testing scenario:**

1.  Configure phpmd
2. Remove `@SuppressWarnings(PHPMD.UnusedFormalParameter)` from some plugin
3. Add `PluginInterface` implementation into the class
4. Run `phpmd`. Expected no `UnusedFormalParameter` errors


**Before:** :heavy_multiplication_x: No information of the invoice email was shown in the header.
Screenshot 2023-08-24 at 9 03 55 AM

**After:** :heavy_check_mark:  Mail sent information was now getting displayed.
Screenshot 2023-08-24 at 9 11 50 AM

Builds are failed. Hence, moving this PR to **Extended Testing.**

**Comment by engcom-Echo** [comment]
@magento run Unit Tests,Functional Tests CE,Functional Tests EE,Functional Tests B2B

**Comment by engcom-Echo** [comment]
@magento run Unit Tests,Functional Tests EE,Functional Tests CE ,Functional Tests B2B

**Comment by engcom-Echo** [comment]
Functional Tests B2B failure is different on last two run on same commit.
Run1
Screenshot 2023-08-25 at 10 14 05 AM

Run2
Screenshot 2023-08-25 at 10 14 11 AM

Hence moving it to `Merge In Progress`.

**Pull request merged**
